### PR TITLE
[SUP-640] Fix bug in GPU selection

### PR DESCRIPTION
**Pull request by kyuksel**
[SUP-640](https://broadworkbench.atlassian.net/browse/SUP-640)

Steps to reproduce on `dev`: Follow the instructions on the ticket above or the screenrecording below.

When a high number of CPUs (e.g. `64`) was selected, the modal's display of `GPUs` got correctly changed to a valid number since Google Cloud doesn't allow for creating a runtime with a high number of CPUs and low number of GPUs. The issue was that, the modal's state wasn't being updated for that new number of GPUs so when a user clicked `CREATE`, the API request contained the stale (low) number of GPUs. The screenrecording below was taken with the fix where one can see that the API request has the correct number of GPUs (`4`).

Tested by flipping through `CPUs`, `Memory`, `GPU type` and `GPUs` options and observing that the corresponding state properties changed as expected.

https://user-images.githubusercontent.com/5438223/162498132-a5cf3ad6-e8fd-46be-bffa-af801520a321.mov

### CODE REVIEW

File: src/components/ComputeModal.js [hunk 1]
```diff
@@ -182,7 +182,9 @@ const shouldUsePersistentDisk = (runtimeType, runtimeDetails, upgradeDiskSelecte
   (!runtimeDetails?.runtimeConfig?.diskSize || upgradeDiskSelected)
 // Auxiliary functions -- end
 
-export const ComputeModalBase = ({ onDismiss, onSuccess, runtimes, persistentDisks, tool, workspace, location, isAnalysisMode = false, shouldHideCloseButton = isAnalysisMode }) => {
+export const ComputeModalBase = ({
+  onDismiss, onSuccess, runtimes, persistentDisks, tool, workspace, location, isAnalysisMode = false, shouldHideCloseButton = isAnalysisMode
+}) => {
```
Comment (kyuksel) [inline src/components/ComputeModal.js:187]: Reformatting only

File: src/components/ComputeModal.js [hunk 2]
```diff
@@ -328,6 +330,24 @@ export const ComputeModalBase = ({ onDismiss, onSuccess, runtimes, persistentDis
...
+    const validGpuName = _.includes(displayNameForGpuType(computeConfig.gpuType), validGpuNames) ?
+      displayNameForGpuType(computeConfig.gpuType) :
+      _.head(validGpuNames)
+    const validGpuType = _.find({ name: validGpuName }, validGpuOptions)?.type
+    const validNumGpusOptions = _.flow(_.filter({ name: validGpuName }), _.map('numGpus'))(validGpuOptions)
+    const validNumGpus = _.includes(computeConfig.numGpus, validNumGpusOptions) ? computeConfig.numGpus : _.head(validNumGpusOptions)
+
+    return { currentNumCpus, currentMemory, validGpuName, validGpuNames, validGpuType, validGpuOptions, validNumGpus, validNumGpusOptions }
+  }
+
```
Comment (kyuksel) [inline src/components/ComputeModal.js:350]: These now need to be computed in multiple places so I factored out the code that used to be on top of `renderComputeProfileSection` below and made it a helper function.

File: src/components/ComputeModal.js [hunk 3]
```diff
@@ -899,8 +914,13 @@ export const ComputeModalBase = ({ onDismiss, onSuccess, runtimes, persistentDis
...
                   isSearchable: false,
                   value: currentNumCpus,
-                  onChange: ({ value }) => updateComputeConfig('masterMachineType',
-                    _.find({ cpu: value }, validMachineTypes)?.name || mainMachineType),
+                  onChange: ({ value }) => {
+                    const mainMachineType = getMainMachineTypeByNumCpus(value)
+                    const { validGpuType: newGpuType, validNumGpus: newNumGpus } = getValidCpuGpuConfig(mainMachineType)
+                    updateComputeConfig('masterMachineType', mainMachineType)
+                    updateComputeConfig('gpuType', newGpuType)
+                    updateComputeConfig('numGpus', newNumGpus)
```
Comment (kyuksel) [inline src/components/ComputeModal.js:922]: These two lines are what fix the bug where we weren't updating the GPU config accordingly to the newly selected machine type. Similarly for the other `Select` below.

Comment (kyuksel) [reply src/components/ComputeModal.js:922]: Considered creating a composite out of `gpuType` and `numGpus` but there are two places where they don't get updated at the same time so decided to keep the current way.

File: src/libs/runtime-utils.js [hunk 4]
```diff
@@ -92,11 +92,11 @@ export const getValidGpuTypesForZone = zone => {
...
 
-export const getValidGpuTypes = (numCpus, mem, zone) => {
-  const validGpuTypesForZone = getValidGpuTypesForZone(zone)
-  const validGpuTypes = _.filter(({ maxNumCpus, maxMem, type }) => numCpus <= maxNumCpus && mem <= maxMem && validGpuTypesForZone.includes(type),
+export const getValidGpuOptions = (numCpus, mem, zone) => {
+  const validGpuOptionsForZone = getValidGpuTypesForZone(zone)
+  const validGpuOptions = _.filter(({ maxNumCpus, maxMem, type }) => numCpus <= maxNumCpus && mem <= maxMem && validGpuOptionsForZone.includes(type),
     gpuTypes)
-  return validGpuTypes || { name: '?', type: '?', numGpus: '?', maxNumCpus: '?', maxMem: '?', price: NaN, preemptiblePrice: NaN }
+  return validGpuOptions || { name: '?', type: '?', numGpus: '?', maxNumCpus: '?', maxMem: '?', price: NaN, preemptiblePrice: NaN }
```
Comment (kyuksel) [inline src/libs/runtime-utils.js:99]: Renaming only to better reflect the values

**Review by Qi77Qi — APPROVED** [review]
thanks @kyuksel ! the recording looks great!

File: src/components/ComputeModal.js [hunk 5]
```diff
@@ -26,7 +26,7 @@ import {
   computeStyles, defaultAutopauseThreshold, defaultComputeRegion, defaultComputeZone, defaultDataprocMachineType, defaultDataprocMasterDiskSize,
   defaultDataprocWorkerDiskSize, defaultGceBootDiskSize, defaultGceMachineType, defaultGcePersistentDiskSize, defaultGpuType, defaultLocation,
   defaultNumDataprocPreemptibleWorkers, defaultNumDataprocWorkers, defaultNumGpus, displayNameForGpuType, findMachineType, getAutopauseThreshold,
-  getCurrentRuntime, getDefaultMachineType, getIsRuntimeBusy, getPersistentDiskCostMonthly, getValidGpuTypes, getValidGpuTypesForZone,
+  getCurrentRuntime, getDefaultMachineType, getIsRuntimeBusy, getPersistentDiskCostMonthly, getValidGpuOptions, getValidGpuTypesForZone,
```
Comment (kyuksel) [inline src/components/ComputeModal.js:29]:
```suggestion
  getCurrentRuntime,getIsRuntimeBusy, getPersistentDiskCostMonthly, getValidGpuOptions, getValidGpuTypesForZone,
```

**Review by dmohs — APPROVED** [review]
Inline PR comments were very helpful, thank you.

**Pull request merged**